cpp/src/Ice/IdleTimeoutTransceiverDecorator.cpp

```diff
@@ -0,0 +1,180 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+#include "IdleTimeoutTransceiverDecorator.h"
+#include "Ice/Buffer.h"
+
+#include <chrono>
+
+using namespace std;
+using namespace Ice;
+using namespace IceInternal;
+
+namespace
+{
+    class HeartbeatTimerTask final : public IceUtil::TimerTask
+    {
+    public:
+        HeartbeatTimerTask(const ConnectionIPtr& connection) : _connection(connection) {}
+
+        void runTimerTask() final
+        {
+            if (auto connection = _connection.lock())
+            {
+                connection->sendHeartbeat();
+            }
+            // else nothing to do, the connection is already gone.
+        }
+
+    private:
+        const weak_ptr<ConnectionI> _connection;
+    };
+
+    class IdleCheckTimerTask final : public IceUtil::TimerTask, public enable_shared_from_this<IdleCheckTimerTask>
+    {
+    public:
+        IdleCheckTimerTask(const ConnectionIPtr& connection, const chrono::milliseconds& idleTimeout)
+            : _connection(connection),
+              _idleTimeout(idleTimeout)
+        {
+        }
+
+        void runTimerTask() final
+        {
+            if (auto connection = _connection.lock())
+            {
+                connection->idleCheck(shared_from_this(), _idleTimeout);
+            }
+            // else nothing to do, the connection is already gone.
+        }
+
+    private:
+        const weak_ptr<ConnectionI> _connection;
+        const chrono::milliseconds _idleTimeout;
+    };
+}
+
+void
+IdleTimeoutTransceiverDecorator::decoratorInit(const ConnectionIPtr& connection)
+{
+    _heartbeatTimerTask = make_shared<HeartbeatTimerTask>(connection);
+    if (_enableIdleCheck)
+    {
+        _idleCheckTimerTask = make_shared<IdleCheckTimerTask>(connection, _idleTimeout);
+    }
+}
+
+SocketOperation
+IdleTimeoutTransceiverDecorator::initialize(Buffer& readBuffer, Buffer& writeBuffer)
+{
+    SocketOperation op = _decoratee->initialize(readBuffer, writeBuffer);
+
+    if (op == SocketOperationNone) // connected
+    {
+        // cancelPrevious = true because Ice often writes to a client connection before it's connected.
+        _timer->schedule(_heartbeatTimerTask, _idleTimeout / 2, true);
+        if (_enableIdleCheck)
+        {
+            // cancelPrevious = true because with SSL, the connection is connected after a read.
+            _timer->schedule(_idleCheckTimerTask, _idleTimeout, true);
+        }
+    }
+
+    return op;
+}
+
+IdleTimeoutTransceiverDecorator::~IdleTimeoutTransceiverDecorator()
+{
+    // If we destroy this object before calling decoratorInit(), _heartbeatTimerTask and _idleCheckTimerTask will be
```

Comment:
This would happen with ConnectionI's constructor throws an exception.

Unless we're out of memory, ConnectionI's ctor should not throw. Do you think I should make ConnectionI's ctor noexcept and add an assert in this decorator dtor?

Reply:
We don't handle catostrophic failures such as OOM errors elsewhere so handling it here is a bit surprising. Up to you but I think it's better to not handle it only here.

Reply:
Ok, make ConnectionI's ctor noexcept and added assert here.